lib/ruby_lsp/listeners/document_highlight.rb

```diff
@@ -550,6 +683,34 @@ def node_value(node)
           node.constant_path.slice
         end
       end
+
+      sig { params(node: Prism::Node).returns(T.nilable(Prism::Location)) }
+      def node_closing_keyword_loc(node)
+        location = (
+          case node
+          when Prism::ModuleNode, Prism::ClassNode, Prism::SingletonClassNode, Prism::DefNode, Prism::CaseNode,
+            Prism::IfNode, Prism::UnlessNode, Prism::ForNode
+            node.end_keyword_loc
+          when Prism::WhileNode, Prism::UntilNode
+            node.closing_loc
+          end
+        )
+        return if location.nil?
+
+        # Sometimes, the closing keyword may not be present.
+        location.length.positive? ? location : nil
+      end
```

Comment:
This method is always invoked from places where we already know the node types, which is a bit odd.

Instead of doing that, let's invoke either `end_key_loc` or `closing_loc` directly on the event handlers and let's make the method extracted to add the highlights handle the possibility of the location being `nil`. For example, something like this

```ruby
def on_until_node_enter(node)
  if @target.is_a?(Prism::UntilNode)
    add_matching_end_highlights(node.keyword_loc, node.closing_loc)
  end  
end

def add_matching_end_highlights(keyword_loc, end_loc)
  return unless keyword_loc && end_loc
  
  return unless (covers_target_position?(until_keyword_loc) || covers_target_position?(end_keyword_loc))

  add_highlight(Constant::DocumentHighlightKind::TEXT, keyword_loc)
  add_highlight(Constant::DocumentHighlightKind::TEXT, end_loc)
end
```